org.eclipse.lsp4e/src/org/eclipse/lsp4e/DocumentContentSynchronizer.java

```diff
@@ -72,11 +68,11 @@ final class DocumentContentSynchronizer implements IDocumentListener {
 	private int version = 0;
 	private DidChangeTextDocumentParams changeParams;
 	private long modificationStamp;
-	private final AtomicReference<CompletableFuture<LanguageServer>> lastChangeFuture;
-	private LanguageServer languageServer;
+
 	private IPreferenceStore store;
 
 	public DocumentContentSynchronizer(@NonNull LanguageServerWrapper languageServerWrapper,
+			@NonNull LanguageServer languageServer,
```

Comment:
The `LanguageServerWrapper` already has a reference to the `languageServer`, so why pass it on its own? Could we not add a method to the `LanguageServerWrapper` if we do not want to call `languageServerWrapper.getInitializedServer()` (is it not just fine to do so, or is the problem that it can side effects?)

Reply:
Yeah it's a bit ugly. The key thing I'm trying to achieve with the new dispatch logic is to make it possible to have an (internal) dispatch function of the form getServer().thenComposeAsync({code}, dispatcher), with getServer() doing returning a completable future that guarantees that all the initialisation prerequisites will have completed.

The key thing is that we want in the majority of cases getServer() to return a completable future that has already completed, in which case the block `{code}` will be scheduled straight onto the dispatcher thread, i.e. *now* from the point of view of the caller. This is critical in making sure most of the client code in LSP4e, which is usually initiated from the UI thread, gets its work scheduled at consistent point in time with respect to any document updates.

I think in this case you're right and it's OK to call `getInitializedServer()`. The reason I've done this is if you look at the place where this is constructed in `LanguageServerWrapper`, it's off the back of an initialisation future, so at that point, we are guaranteed to have the `LanguageServer` ready.

I was just a bit over-paranoid about calling some of the methods that return async objects, just in case I accidentally introduced another possible race condition. It's important that the `didOpen` message gets sent to the language server before the object returned by `LanguageServerWrapper.connect()` gets used for any more calls.

https://github.com/eclipse/lsp4e/blob/ad55abb3d56d601d6209ff3f62655929107e49b1/org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java#L673

and https://github.com/eclipse/lsp4e/blob/ad55abb3d56d601d6209ff3f62655929107e49b1/org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java#L658
are critical.

Reply:
I understand that and I think it is good, since you want to be sure ;). I think than better than just passing the LS, would be to have a method that gets you the `languageServer` as is without doing anything else. Given that `getServer` already exists but calls `getInitializedServer` internally, I would propose we create a new method `unwrap` in LSW that just `return languageServer;`. What do you think of that?